dipy/reconst/ivim.py

```diff
@@ -364,20 +368,28 @@ def estimate_linear_fit(self, data, split_b, less_than=True):
 
         D : float
             The estimated value of D.
+
+        References
+        ----------
+        Lawson C., Hanson R.J., (1987) Solving Least Squares Problems, SIAM
+
         """
         if less_than:
             bvals_split = self.gtab.bvals[self.gtab.bvals <= split_b]
-            D, neg_log_S0 = np.polyfit(bvals_split,
-                                       -np.log(data[self.gtab.bvals <=
-                                                    split_b]), 1)
+            b = -np.log(data[self.gtab.bvals <= split_b])
+
+            D = nnls(np.matrix(bvals_split).T, b)[0]
```

Comment:
The variable name `b` is a bit confusing here. Maybe `neg_log_S`? 

Also, why the negative sign?

Finally, I might be confused: to fit diffusivity, shouldn't we be using the data from b-values larger than `split_b`?

Reply:
For the >= `split_b` see line `385`

and I am using the negative sign as it was in the original implementation, therefore did not touch it...

I will fix the variable names :+1:

Reply:
Does `nnls` require the input to be a numpy `matrix` object? If not, I would avoid using that class -- it's deprecated (see https://stackoverflow.com/a/53254739/3532933 for a lot of detail about this).

Reply:
Yes, its compulsory! Throws an error otherwise... I tried it.

Reply:
Also, one more thought: do we need to also include an intercept term here? Maybe add a column of 1's to the design matrix?

Reply:
I think that I see why. If you do this:

```
X = np.random.randn(100)                                               
y = np.random.randn(100)                                               
coef = nnls(X, y)   
```

it raises:

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-14-1a1020100ad2> in <module>
----> 1 coef = nnls(X, y)

~/.virtualenvs/dipy/lib/python3.7/site-packages/scipy/optimize/nnls.py in nnls(A, b, maxiter)
     44 
     45     if len(A.shape) != 2:
---> 46         raise ValueError("expected matrix")
     47     if len(b.shape) != 1:
     48         raise ValueError("expected vector")

ValueError: expected matrix

```

But it's not a great error message, because it suggests that you need a `np.matrix` object, where really what is required is an array with more two dimensions (see line 45 in that traceback), so this works: 

```
coef = nnls(X[:, np.newaxis], y)
```

without using a `np.matrix` object at all. This will also work when you add ones to account for the intercept. Something like: 

```
coef = nnls(np.vstack([X, np.ones(X.shape[0])]).T, y)    
```

should work for that.

Reply:
See https://github.com/scipy/scipy/pull/11574

Reply:
Thank you for this @arokem ! Let me try this and get back ASAP :+1:

Reply:
![image](https://user-images.githubusercontent.com/17598557/75207686-737ba900-5747-11ea-824b-28bd86db322d.png)
This is what the `D` map looks like. The black region is mostly zero. I tried both of the above suggestions made by you :+1: still don't see any difference...

Even the D* looks similar:
![image](https://user-images.githubusercontent.com/17598557/75207760-a887fb80-5747-11ea-95f1-10177d240fa7.png)

Maybe the results aren't wrong, but I haven't seen exactly `0` values for `D` and `D*`.

Reply:
@arokem, one more thing I forgot to mention is that `NNLS` seems to become sensitive to the bounds. For instance, when I change the bounds to `([0, 0, 0, 0], [np.inf, .9, .9, .9])` the result for D looks like this:
![image](https://user-images.githubusercontent.com/17598557/75266536-7bc3fa80-57c0-11ea-97a5-71a078cfbc91.png)

The places where you see the black spots above are perfect `0` or something like `2.9e-28`

Reply:
That's odd. As far as I can tell, NNLS doesn't know anything about these bounds. Unless I am missing something?

Reply:
Nope.. I think you are right! But I also think that the results that we are seeing are not wrong either. Think of a case where there is a lot of noise, if you solve a QP, it is likely that we see a horizontal line right?

I was looking into what others are doing, and seems like they do a `polyfit` and discard+interpolate the values where D is negative. For instance: https://www.mathworks.com/matlabcentral/fileexchange/65579-ivim-model-fitting -> `IVIM_seg`

Does this: 
![image](https://user-images.githubusercontent.com/17598557/75287069-b725ef80-57e7-11ea-9026-544693f65093.png)

Reply:
Hey @ShreyasFadnavis : could you try changing this line to  `coef = nnls(np.vstack([X, np.ones(X.shape[0])]).T, y)` as suggested above?

Reply:
Hi @arokem I did try doing this.. but did not give anything better! I just did not update this PR with that commit...